### DISCUSSION TITLE
ein diff ?

### DISCUSSION THREAD

**Discussion by oknozor** [GitoxideLabs/gitoxide#657] [Ideas]
Today I worked on git-diff in my project which is the last feature needed to implement pull request. 

I played a bit with gitoxide and the [diffs](https://crates.io/crates/diffs) crate (maintained by the pijul folks). And was able to get a simple diff output. 

It's far from finished and need some polish but, it is working.

Let me know if this is something that you would want in ein. 

![screenshot-2022-12-13T20:57:26](https://user-images.githubusercontent.com/24563836/207432118-700ad980-d20c-47aa-929d-4748e47d5ec1.png)

https://github.com/oknozor/gill/blob/testdiff/crates/gill-git/src/diff.rs
https://github.com/oknozor/gill/blob/testdiff/crates/gill-git/src/repository/diff.rs

**Comment by Byron** [comment]
Thanks for thinking of me!

The driver for all textual diffs that is also geared towards being as good as `git` or better is `imara-diff`, and it's now imported verbatim in the `git-diff` crate. Its author, @pascalkuthe, is planning to integrate a diff visualization into `gix diff` when he gets a chance.

That said, `ein` is probably going to receive diff functionality one day, but it will be tightly integrated into the user experience, whatever that may look like. That said, I will definitely keep `gill` in mind when the time comes.

Besides that, it's exciting to see you are building what appears to be a federated GitHub-like system. I hope soon you can use `gitoxide` for `upload-pack` as well, along with an extensible `git` server.

**Comment by ghost** [comment]
so is diff command currently available or not? I tried these:

~~~
> gix diff
error: unrecognized subcommand 'diff'

> ein diff
error: unrecognized subcommand 'diff'
~~~

build command:

~~~
cargo build --no-default-features --features max-pure
~~~

**Reply by Byron** [reply]
No, diffing isn't yet available via `gix`. The functionality to diff trees of commits and lines of blobs is available though, so a tool like `gix diff` can already be implemented. If it was implemented as `gix diff`, it would be a proof of concept as it probably wouldn't be better necessarily than [`delta`](https://github.com/dandavison/delta) for example.

**Reply by EliahKagan** [reply]
The functionality of `gix` in this area is substantially expanded since 2023, but not necessarily to the point that you would need. The good news is that these days there is a `gix diff` subcommand. It is used through its own subcommands `tree` and `file` for diffing trees and files, respectively:

```text
ek@Cup ~/r/gitoxide (main| =|✔)> gix diff
Print all changes between two objects

Usage: gix diff <COMMAND>

Commands:
  tree  Diff two trees
  file  Diff two versions of a file
  help  Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help
```

For example, diffing a recent change:

```text
ek@Cup ~/r/gitoxide (main| =|✔)> gix diff tree HEAD^ HEAD
Diffing trees `HEAD^` (f7c7145df0f069b58d1e64876ba21b273f52a427) -> `HEAD` (f8be65fef6ed3cdc55de531da30d0b4ad0ce0f14)

M: Cargo.toml
  fc491700a -> 69375522a
ek@Cup ~/r/gitoxide (main| =|✔)> gix diff file HEAD^:Cargo.toml HEAD:Cargo.toml
@@ -58,7 +58,7 @@
 
 ## The smallest possible build, best suitable for small single-core machines.
 ##
-## This build is essentially limited to local operations without any fanciness.
+## This build is essentially limited to local operations without any fanciness. It does not have `gix clone`.
 ##
 ## Optimized for size, no parallelism thus much slower, progress line rendering.
 small = ["pretty-cli", "prodash-render-line", "is-terminal"]
```

However, I don't know if this would meet your needs or not. This command does not have all the functionality of `git diff`. In particular, it does not interact with the index, which is a very important part of regular `git diff` functionality for most Git users.

**Reply by Byron** [reply]
I'd recommend against using `gix` as driver for anything as it's a development tool to help *test* `gitoxide`, and nothing more.
`ein` would be the tool to go for as it aims to be the Git we always wanted, but it's unclear when this will happen, if ever.

Right now all my time goes into [GitButler](https://gitbutler.com), which ships a command-line application as well these days. It's likely to get diffing first, and it *should* be geared to eventually be able to replace Git with a simpler workflow.

**Reply by 3052** [reply]
thanks but I am old school I want command line only - looks nice though

**Reply by Byron** [reply]
It comes with a command-line tool, and if it gets diffs, they will be better than what `gix diff` offers for sure.

**Reply by 3052** [reply]
looks like the command line tool cant even clone, so it might be premature to recommend it as a C Git alternative

**Reply by Byron** [reply]
`gix` can clone, `ein` cannot. If you can point out where `gix` is even implied to be a Git command-line alternative, I'll happily fix it.

**Reply by 3052** [reply]
I am talking about GitButler. you made it as a suggestion but its cant even clone
